python/ray/data/_internal/util.py

```diff
@@ -51,3 +58,97 @@ def _check_pyarrow_version():
             )
         else:
             _VERSION_VALIDATED = True
+
+
+def _autodetect_parallelism(
+    parallelism: int,
+    cur_pg: Optional["PlacementGroup"],
+    ctx: DatasetContext,
+    reader: Optional["Reader"] = None,
+    avail_cpus: Optional[int] = None,
+) -> (int, int):
+    """Returns parallelism to use and the min safe parallelism to avoid OOMs.
+
+    This detects parallelism based on the following heuristics (highest to lowest
+    precedence):
+
+     1) Available CPUs. We detect how many CPUs are available for use in the cluster,
+        and ensure the parallelism is high enough to make use of the CPUs.
+     2) Min block size. We try to avoid creating blocks smaller than this threshold,
+        to avoid the overhead of tiny blocks.
+     3) Max block size. We try to avoid creating blocks larger than this threshold,
+        since it can lead to OOM errors during processing.
+     4) If none of the above apply, the default parallelism of 200 is used.
+
+    Args:
+        parallelism: The user-requested parallelism, or -1 for auto-detection.
+        cur_pg: The current placement group, to be used for avail cpu calculation.
+        ctx: The current Dataset context to use for configs.
+        reader: The datasource reader, to be used for data size estimation.
+        avail_cpus: Override avail cpus detection (for testing only).
+
+    Returns:
+        Tuple of detected parallelism (only if -1 was specified), and the min safe
+        parallelism (which can be used to generate warnings about large blocks).
+    """
+    min_safe_parallelism = 1
+    if reader:
+        mem_size = reader.estimate_inmemory_data_size()
+        if mem_size is not None:
+            min_safe_parallelism = max(1, int(mem_size / ctx.target_max_block_size))
+            max_reasonable_parallelism = max(
```

Comment:
are these max() needed since the only case it applies is when our target max / min block size is greater than mem_size, which seems we should just throw ?

Reply:
We should never throw; those are signals to adjust the parallelism to meet the requirements.